### ISSUE TITLE
Confusing behavior for default providers

### ISSUE THREAD

**Issue by mrpatrickpilch** [keycloak/keycloak#16626]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

dist/quarkus

### Describe the bug

When trying to create an _additional_ custom `LoginFormsProvider` to work side by side with a default, login forms are completely broken.

### Version

20.0.1

### Expected behavior

My intuition would say that the _**default**_ provider would be used in the case no explicit provider is requested. E.g. `KeycloakSession#getProvider(Class<T> clazz)` vs `KeycloakSession#getProvider(Class<T> clazz, String id)`.

Example:
- Create `MyCustomLoginFormsProvider extends LoginFormsProvider` whose factory returns "mycustomloginforms" as the ID.
- Launch KC with `--spi-login-provider=mycustomloginforms`.

**I would expect the following:**

`keycloakSession.getProvider(LoginFormsProvider.class)` to return `MyCustomLoginFormsProvider`.

`keycloakSession.getProvider(LoginFormsProvider.class, "freemarker")` to return the out-of-the-box  `org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider`.

### Actual behavior

Keycloak's behavior is very surprising to me. When a "default" provider is defined, it appears to actually be the "sole" provider.

In the example above, the factory for `org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider` is _never loaded._

### How to Reproduce?

See example. Can provide reproduction provider JAR if requested.

### Anything else?

As of 20.0.3, the [provider factory loading logic](https://github.com/keycloak/keycloak/blob/20.0.3/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java#L625-L635) will ignore all other factories if a "default" is specified.

```java
String provider = Config.getProvider(spi.getName());

if (provider == null) {
    loadedFactories.addAll(pm.load(spi));
} else {
    ProviderFactory factory = pm.load(spi, provider);

    if (factory != null) {
        loadedFactories.add(factory);
    }
}
```

**Comment by mrpatrickpilch** [comment]
Sorry, I just found another similar issue after the fact. https://github.com/keycloak/keycloak/issues/11239

The suggestion https://github.com/keycloak/keycloak/issues/11239#issuecomment-1138690834 proposed in that thread does not work in this case. 

`FreeMarkerLoginFormsProviderFactory` has order 0. If I make my custom provider < 0, it fails [this conditional check](https://github.com/keycloak/keycloak/blob/20.0.3/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java#L782) 

But in this case, setting the custom provider's ID to "default" achieves what I want. 

In summary, I still feel this behavior is confusing and has rough edges. If improvements can't be made, then perhaps the server development documentation could include more detail at least.

**Comment by pedroigor** [comment]
@mrpatrickpilch I'm not sure if this is an issue. IIRC, it is also how legacy distribution works.

You can set a default and that means not loading others in order to reduce the number of installed providers.

Is this doc enough https://www.keycloak.org/server/configuration-provider#_configuring_a_default_provider?

**Comment by mrpatrickpilch** [comment]
@pedroigor Hi, thanks for the reply. I just want to share my opinion about why I do consider this an issue. One of my reasons would be a violation of the principle of least astonishment, based on me having sunk this many hours into this and the 2 referenced issues others have stumbled with as well. But I'll concede that this ticket may not be a defect/regression, but perhaps more of a suggestion/improvement at this point.

> You can set a default and that means not loading others in order to reduce the number of installed providers.

I am now aware :) But to be honest, I don't find this functionality particularly useful because I have the option to 1) remove the provider from the providers directory, or 2) [just disabling it](https://www.keycloak.org/server/configuration-provider#_enabling_and_disabling_a_provider) if I find a loaded provider to be problematic. I have not run into, or can even think of, a situation in which I would require this. I think there's obviously a good case for configuring a true _default_ provider among 2 or more installed providers.

> Is this doc enough https://www.keycloak.org/server/configuration-provider#_configuring_a_default_provider?

I must admit: reading this, I already had a preconceived notion of how this was "supposed" to work. But I still feel it's not un-ambiguously clear. 

> "Depending on the SPI, multiple provider implementations can co-exist but only one of them is going to be used at runtime. For these SPIs, a default provider is the primary implementation that is going to be active and used at runtime."

So are there really multiple _co-existing_ provider implementations? I say _no_. In this existing "default" behavior, there is only _one_ provider. The others are just ignored jar files on the file system, completely irrelevant to the issue at hand. I really thought the reason Keycloak's API separated `KeycloakSession#getProvider(Class<T> clazz)` vs `KeycloakSession#getProvider(Class<T> clazz, String id)` was in order to allow _multiple_ providers to be loaded with the first API giving you the "default", and the latter letting you be specific about which you'd like. After having dedicated lots of time investigating this, I am now aware there is entirely different "default" behavior for the SPI CLI arguments, and the provider factory ID of "default". Kind of confusing, no?

And I do understand there are the options of `org.keycloak.provider.ProviderFactory#order` and an ID of "default". But what I tried to illustrate in my comments is that:
1. `#order` is problematic due to the < 0 condition. It was not helpful in my particular situation.
2. An id of "default" is likely to be problematic. For now, I've worked around the issue. But tomorrow, for example, I would be in a really difficult situation if I had to load a 3rd party provider that also returned an ID of "default" to address a similar issue to mine.

**Comment by pedroigor** [comment]
Thank you for the feedback.

> I am now aware :) But to be honest, I don't find this functionality particularly useful because I have the option to 1) remove the provider from the providers directory, or 2) [just disabling it](https://www.keycloak.org/server/configuration-provider#_enabling_and_disabling_a_provider) if I find a loaded provider to be problematic. I have not run into, or can even think of, a situation in which I would require this. I think there's obviously a good case for configuring a true default provider among 2 or more installed providers.

The reason behind that functionality is to make the server more opinionated about the runtime and reduce memory footprint when you want a single provider active. I agree you can achieve the same by individually removing providers but there are built-in providers that you don't need to know their ids but just replace with yours.

The value `default` for a factory ID is a convention that when you don't explicitly set a provider as the default, the factory with an id `default` will be used. Although not documented, only built-in providers should be using `default` as an ID, not custom providers.

That said, I think we should update our docs and be more clear about how providers are registered and removed from the runtime as well as explain better how the built-in providers using the `default` ID work and that they this value is reserved not usable by custom providers.

Does it make sense to you?

**Comment by mrpatrickpilch** [comment]
I am in favor of the documentation update you propose. Thank you for the clarification and time.

**Comment by william-will-angi** [comment]
Am I understanding correctly here that there is only the possibility for a single forms provider at a time?
